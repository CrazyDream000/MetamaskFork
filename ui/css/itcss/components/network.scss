.network-component--disabled {
  cursor: not-allowed;

  .fa-caret-down {
    opacity: 0;
  }
}

.network-component.pointer {
  border: 1px solid $Grey-200;
  border-radius: 82px;
  padding: 6px 3px;
  flex: 0 0 auto;
  display: flex;

  &.ethereum-network .menu-icon-circle div {
    background-color: rgba(3, 135, 137, 0.7) !important;
  }

  &.ropsten-test-network .menu-icon-circle div {
    background-color: rgba(233, 21, 80, 0.7) !important;
  }

  &.kovan-test-network .menu-icon-circle div {
    background-color: rgba(105, 4, 150, 0.7) !important;
  }

  &.rinkeby-test-network .menu-icon-circle div {
    background-color: rgba(235, 179, 63, 0.7) !important;
  }

  &.goerli-test-network .menu-icon-circle div {
    background-color: rgba(48, 153, 242, 0.7) !important;
  }
}

.dropdown-menu-item {
  .menu-icon-circle,
  .menu-icon-circle--active {
    margin: 0 14px;
  }
}

.dropdown-menu-item .fa.delete {
  margin-right: 10px;
  display: none;
}

.dropdown-menu-item:hover .fa.delete {
  display: inherit;
}

.network-droppo {
  right: 2px;

  .color-indicator {
    margin: 0 14px;
  }

  @media screen and (min-width: 576px) {
    right: calc(((100% - 85vw) / 2) + 2px);
  }

  @media screen and (min-width: 769px) {
    right: calc(((100% - 80vw) / 2) + 2px);
  }

  @media screen and (min-width: 1281px) {
    right: calc(((100% - 65vw) / 2) + 2px);
  }
}

.network-name-item {
  flex: 1;
  color: $dusty-gray;
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

.network-check,
.network-check__transparent {
  color: $white;
  margin-left: 7px;
}

.network-check__transparent {
  opacity: 0;
  width: 16px;
  margin: 0;
}

.menu-icon-circle,
.menu-icon-circle--active {
  background: none;
  border-radius: 22px;
  display: flex;
  justify-content: center;
  align-items: center;
  border: 1px solid transparent;
  margin: 0 4px;
}

.menu-icon-circle--active {
  border: 1px solid $white;
  background: rgba(100, 100, 100, 0.4);
}

.menu-icon-circle div,
.menu-icon-circle--active div {
  height: 12px;
  width: 12px;
  border-radius: 17px;
}

.menu-icon-circle--active div {
  opacity: 1;
}

.network-dropdown-header {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 100%;
}

.network-dropdown-divider {
  width: 100%;
  height: 1px;
  margin: 10px 0;
  background-color: $scorpion;
}

.network-dropdown-title {
  @include H4;

  height: 25px;
  width: 120px;
  color: $white;
  text-align: center;
}

.network-dropdown-content {
  @include H6;

  min-height: 36px;
  width: 265px;
  color: $dusty-gray;
}

.network-caret {
  margin: 0 8px;
}

.network-loading-spinner {
  display: flex;
  flex-flow: row nowrap;
  align-items: center;
  position: relative;
  height: 16px;
  width: 16px;
  margin-left: 5px;

  img {
    height: 26px;
    position: absolute;
    top: -5px;
    left: -6px;
  }
}
