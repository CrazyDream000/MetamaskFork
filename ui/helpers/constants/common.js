export const ETH = 'ETH';
export const GWEI = 'GWEI';
export const WEI = 'WEI';

export const PRIMARY = 'PRIMARY';
export const SECONDARY = 'SECONDARY';

export const ERC20 = 'ERC20';

export const GAS_ESTIMATE_TYPES = {
  SLOW: 'SLOW',
  AVERAGE: 'AVERAGE',
  FAST: 'FAST',
  FASTEST: 'FASTEST',
};
