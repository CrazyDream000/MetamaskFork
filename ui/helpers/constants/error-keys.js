export const INSUFFICIENT_FUNDS_ERROR_KEY = 'insufficientFunds';
export const GAS_LIMIT_TOO_LOW_ERROR_KEY = 'gasLimitTooLow';
export const TRANSACTION_ERROR_KEY = 'transactionError';
export const TRANSACTION_NO_CONTRACT_ERROR_KEY = 'transactionErrorNoContract';
export const ETH_GAS_PRICE_FETCH_WARNING_KEY = 'ethGasPriceFetchWarning';
export const GAS_PRICE_FETCH_FAILURE_ERROR_KEY = 'gasPriceFetchFailed';
export const GAS_PRICE_EXCESSIVE_ERROR_KEY = 'gasPriceExcessive';
export const UNSENDABLE_ASSET_ERROR_KEY = 'unsendableAsset';
