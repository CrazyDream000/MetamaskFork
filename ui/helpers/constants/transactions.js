import {
  TRANSACTION_TYPES,
  TRANSACTION_STATUSES,
} from '../../../shared/constants/transaction';

export const PENDING_STATUS_HASH = {
  [TRANSACTION_STATUSES.UNAPPROVED]: true,
  [TRANSACTION_STATUSES.APPROVED]: true,
  [TRANSACTION_STATUSES.SUBMITTED]: true,
};

export const PRIORITY_STATUS_HASH = {
  ...PENDING_STATUS_HASH,
  [TRANSACTION_STATUSES.CONFIRMED]: true,
};

export const TOKEN_CATEGORY_HASH = {
  [TRANSACTION_TYPES.TOKEN_METHOD_APPROVE]: true,
  [TRANSACTION_TYPES.TOKEN_METHOD_TRANSFER]: true,
  [TRANSACTION_TYPES.TOKEN_METHOD_TRANSFER_FROM]: true,
};
