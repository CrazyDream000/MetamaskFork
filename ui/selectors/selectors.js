import { stripHexPrefix } from 'ethereumjs-util';
import { createSelector } from 'reselect';
import { addHexPrefix } from '../../app/scripts/lib/util';
import {
  MAINNET_CHAIN_ID,
  BSC_CHAIN_ID,
  TEST_CHAINS,
  NETWORK_TYPE_RPC,
  NATIVE_CURRENCY_TOKEN_IMAGE_MAP,
} from '../../shared/constants/network';
import { KEYRING_TYPES } from '../../shared/constants/hardware-wallets';

import {
  SWAPS_CHAINID_DEFAULT_TOKEN_MAP,
  ALLOWED_SWAPS_CHAIN_IDS,
} from '../../shared/constants/swaps';

import { shortenAddress, getAccountByAddress } from '../helpers/utils/util';
import {
  getValueFromWeiHex,
  hexToDecimal,
} from '../helpers/utils/conversions.util';

import { TEMPLATED_CONFIRMATION_MESSAGE_TYPES } from '../pages/confirmation/templates';

import { toChecksumHexAddress } from '../../shared/modules/hexstring-utils';
import { DAY } from '../../shared/constants/time';
import {
  getNativeCurrency,
  getConversionRate,
  isEIP1559Network,
} from '../ducks/metamask/metamask';

/**
 * One of the only remaining valid uses of selecting the network subkey of the
 * metamask state tree is to determine if the network is currently 'loading'.
 *
 * This will be used for all cases where this state key is accessed only for that
 * purpose.
 * @param {Object} state - redux state object
 */
export function isNetworkLoading(state) {
  return state.metamask.network === 'loading';
}

export function getNetworkIdentifier(state) {
  const {
    metamask: {
      provider: { type, nickname, rpcUrl },
    },
  } = state;

  return nickname || rpcUrl || type;
}

export function getMetricsNetworkIdentifier(state) {
  const { provider } = state.metamask;
  return provider.type === NETWORK_TYPE_RPC ? provider.rpcUrl : provider.type;
}

export function getCurrentChainId(state) {
  const { chainId } = state.metamask.provider;
  return chainId;
}

export function getCurrentKeyring(state) {
  const identity = getSelectedIdentity(state);

  if (!identity) {
    return null;
  }

  const simpleAddress = stripHexPrefix(identity.address).toLowerCase();

  const keyring = state.metamask.keyrings.find((kr) => {
    return (
      kr.accounts.includes(simpleAddress) ||
      kr.accounts.includes(identity.address)
    );
  });

  return keyring;
}

export function isEIP1559Account(state) {
  // Trezor does not support 1559 at this time
  const currentKeyring = getCurrentKeyring(state);
  return currentKeyring && currentKeyring.type !== KEYRING_TYPES.TREZOR;
}

export function checkNetworkAndAccountSupports1559(state) {
  const networkSupports1559 = isEIP1559Network(state);
  const accountSupports1559 = isEIP1559Account(state);

  return networkSupports1559 && accountSupports1559;
}

/**
 * Checks if the current wallet is a hardware wallet.
 * @param {Object} state
 * @returns {Boolean}
 */
export function isHardwareWallet(state) {
  const keyring = getCurrentKeyring(state);
  return Boolean(keyring?.type?.includes('Hardware'));
}

/**
 * Get a HW wallet type, e.g. "Ledger Hardware"
 * @param {Object} state
 * @returns {String|undefined}
 */
export function getHardwareWalletType(state) {
  const keyring = getCurrentKeyring(state);
  return isHardwareWallet(state) ? keyring.type : undefined;
}

export function getAccountType(state) {
  const currentKeyring = getCurrentKeyring(state);
  const type = currentKeyring && currentKeyring.type;

  switch (type) {
    case 'Trezor Hardware':
    case 'Ledger Hardware':
      return 'hardware';
    case 'Simple Key Pair':
      return 'imported';
    default:
      return 'default';
  }
}

/**
 * get the currently selected networkId which will be 'loading' when the
 * network changes. The network id should not be used in most cases,
 * instead use chainId in most situations. There are a limited number of
 * use cases to use this method still, such as when comparing transaction
 * metadata that predates the switch to using chainId.
 * @deprecated - use getCurrentChainId instead
 * @param {Object} state - redux state object
 */
export function deprecatedGetCurrentNetworkId(state) {
  return state.metamask.network;
}

export const getMetaMaskAccounts = createSelector(
  getMetaMaskAccountsRaw,
  getMetaMaskCachedBalances,
  (currentAccounts, cachedBalances) =>
    Object.entries(currentAccounts).reduce(
      (selectedAccounts, [accountID, account]) => {
        if (account.balance === null || account.balance === undefined) {
          return {
            ...selectedAccounts,
            [accountID]: {
              ...account,
              balance: cachedBalances && cachedBalances[accountID],
            },
          };
        }
        return {
          ...selectedAccounts,
          [accountID]: account,
        };
      },
      {},
    ),
);

export function getSelectedAddress(state) {
  return state.metamask.selectedAddress;
}

export function getSelectedIdentity(state) {
  const selectedAddress = getSelectedAddress(state);
  const { identities } = state.metamask;

  return identities[selectedAddress];
}

export function getNumberOfAccounts(state) {
  return Object.keys(state.metamask.accounts).length;
}

export function getNumberOfTokens(state) {
  const { tokens } = state.metamask;
  return tokens ? tokens.length : 0;
}

export function getMetaMaskKeyrings(state) {
  return state.metamask.keyrings;
}

export function getMetaMaskIdentities(state) {
  return state.metamask.identities;
}

export function getMetaMaskAccountsRaw(state) {
  return state.metamask.accounts;
}

export function getMetaMaskCachedBalances(state) {
  const chainId = getCurrentChainId(state);

  // Fallback to fetching cached balances from network id
  // this can eventually be removed
  const network = deprecatedGetCurrentNetworkId(state);

  return (
    state.metamask.cachedBalances[chainId] ??
    state.metamask.cachedBalances[network]
  );
}

/**
 * Get ordered (by keyrings) accounts with identity and balance
 */
export const getMetaMaskAccountsOrdered = createSelector(
  getMetaMaskKeyrings,
  getMetaMaskIdentities,
  getMetaMaskAccounts,
  (keyrings, identities, accounts) =>
    keyrings
      .reduce((list, keyring) => list.concat(keyring.accounts), [])
      .filter((address) => Boolean(identities[address]))
      .map((address) => ({ ...identities[address], ...accounts[address] })),
);

export const getMetaMaskAccountsConnected = createSelector(
  getMetaMaskAccountsOrdered,
  (connectedAccounts) =>
    connectedAccounts.map(({ address }) => address.toLowerCase()),
);

export function isBalanceCached(state) {
  const selectedAccountBalance =
    state.metamask.accounts[getSelectedAddress(state)].balance;
  const cachedBalance = getSelectedAccountCachedBalance(state);

  return Boolean(!selectedAccountBalance && cachedBalance);
}

export function getSelectedAccountCachedBalance(state) {
  const cachedBalances = getMetaMaskCachedBalances(state);
  const selectedAddress = getSelectedAddress(state);

  return cachedBalances && cachedBalances[selectedAddress];
}

export function getSelectedAccount(state) {
  const accounts = getMetaMaskAccounts(state);
  const selectedAddress = getSelectedAddress(state);

  return accounts[selectedAddress];
}

export function getTargetAccount(state, targetAddress) {
  const accounts = getMetaMaskAccounts(state);
  return accounts[targetAddress];
}

export const getTokenExchangeRates = (state) =>
  state.metamask.contractExchangeRates;

export function getAssetImages(state) {
  const assetImages = state.metamask.assetImages || {};
  return assetImages;
}

export function getAddressBook(state) {
  const chainId = getCurrentChainId(state);
  if (!state.metamask.addressBook[chainId]) {
    return [];
  }
  return Object.values(state.metamask.addressBook[chainId]);
}

export function getAddressBookEntry(state, address) {
  const addressBook = getAddressBook(state);
  const entry = addressBook.find(
    (contact) => contact.address === toChecksumHexAddress(address),
  );
  return entry;
}

export function getAddressBookEntryName(state, address) {
  const entry =
    getAddressBookEntry(state, address) || state.metamask.identities[address];
  return entry && entry.name !== '' ? entry.name : shortenAddress(address);
}

export function accountsWithSendEtherInfoSelector(state) {
  const accounts = getMetaMaskAccounts(state);
  const identities = getMetaMaskIdentities(state);

  const accountsWithSendEtherInfo = Object.entries(identities).map(
    ([key, identity]) => {
      return { ...identity, ...accounts[key] };
    },
  );

  return accountsWithSendEtherInfo;
}

export function getAccountsWithLabels(state) {
  return getMetaMaskAccountsOrdered(state).map(
    ({ address, name, balance }) => ({
      address,
      addressLabel: `${name} (...${address.slice(address.length - 4)})`,
      label: name,
      balance,
    }),
  );
}

export function getCurrentAccountWithSendEtherInfo(state) {
  const currentAddress = getSelectedAddress(state);
  const accounts = accountsWithSendEtherInfoSelector(state);

  return getAccountByAddress(accounts, currentAddress);
}

export function getTargetAccountWithSendEtherInfo(state, targetAddress) {
  const accounts = accountsWithSendEtherInfoSelector(state);
  return getAccountByAddress(accounts, targetAddress);
}

export function getCurrentEthBalance(state) {
  return getCurrentAccountWithSendEtherInfo(state).balance;
}

export function getGasIsLoading(state) {
  return state.appState.gasIsLoading;
}

export function getCurrentCurrency(state) {
  return state.metamask.currentCurrency;
}

export function getTotalUnapprovedCount(state) {
  const {
    unapprovedMsgCount = 0,
    unapprovedPersonalMsgCount = 0,
    unapprovedDecryptMsgCount = 0,
    unapprovedEncryptionPublicKeyMsgCount = 0,
    unapprovedTypedMessagesCount = 0,
    pendingApprovalCount = 0,
  } = state.metamask;

  return (
    unapprovedMsgCount +
    unapprovedPersonalMsgCount +
    unapprovedDecryptMsgCount +
    unapprovedEncryptionPublicKeyMsgCount +
    unapprovedTypedMessagesCount +
    getUnapprovedTxCount(state) +
    pendingApprovalCount +
    getSuggestedTokenCount(state)
  );
}

function getUnapprovedTxCount(state) {
  const { unapprovedTxs = {} } = state.metamask;
  return Object.keys(unapprovedTxs).length;
}

export function getUnapprovedConfirmations(state) {
  const { pendingApprovals } = state.metamask;
  return Object.values(pendingApprovals);
}

export function getUnapprovedTemplatedConfirmations(state) {
  const unapprovedConfirmations = getUnapprovedConfirmations(state);
  return unapprovedConfirmations.filter((approval) =>
    TEMPLATED_CONFIRMATION_MESSAGE_TYPES.includes(approval.type),
  );
}

function getSuggestedTokenCount(state) {
  const { suggestedTokens = {} } = state.metamask;
  return Object.keys(suggestedTokens).length;
}

export function getIsMainnet(state) {
  const chainId = getCurrentChainId(state);
  return chainId === MAINNET_CHAIN_ID;
}

export function getIsTestnet(state) {
  const chainId = getCurrentChainId(state);
  return TEST_CHAINS.includes(chainId);
}

export function getIsNonStandardEthChain(state) {
  return !(getIsMainnet(state) || getIsTestnet(state) || process.env.IN_TEST);
}

export function getPreferences({ metamask }) {
  return metamask.preferences;
}

export function getShouldShowFiat(state) {
  const isMainNet = getIsMainnet(state);
  const conversionRate = getConversionRate(state);
  const { showFiatInTestnets } = getPreferences(state);
  return Boolean((isMainNet || showFiatInTestnets) && conversionRate);
}

export function getShouldHideZeroBalanceTokens(state) {
  const { hideZeroBalanceTokens } = getPreferences(state);
  return hideZeroBalanceTokens;
}

export function getAdvancedInlineGasShown(state) {
  return Boolean(state.metamask.featureFlags.advancedInlineGas);
}

export function getUseNonceField(state) {
  return Boolean(state.metamask.useNonceField);
}

export function getCustomNonceValue(state) {
  return String(state.metamask.customNonceValue);
}

export function getDomainMetadata(state) {
  return state.metamask.domainMetadata;
}

export function getRpcPrefsForCurrentProvider(state) {
  const { frequentRpcListDetail, provider } = state.metamask;
  const selectRpcInfo = frequentRpcListDetail.find(
    (rpcInfo) => rpcInfo.rpcUrl === provider.rpcUrl,
  );
  const { rpcPrefs = {} } = selectRpcInfo || {};
  return rpcPrefs;
}

export function getKnownMethodData(state, data) {
  if (!data) {
    return null;
  }
  const prefixedData = addHexPrefix(data);
  const fourBytePrefix = prefixedData.slice(0, 10);
  const { knownMethodData } = state.metamask;

  return knownMethodData && knownMethodData[fourBytePrefix];
}

export function getFeatureFlags(state) {
  return state.metamask.featureFlags;
}

export function getOriginOfCurrentTab(state) {
  return state.activeTab.origin;
}

export function getIpfsGateway(state) {
  return state.metamask.ipfsGateway;
}

export function getInfuraBlocked(state) {
  return Boolean(state.metamask.infuraBlocked);
}

export function getUSDConversionRate(state) {
  return state.metamask.usdConversionRate;
}

export function getWeb3ShimUsageStateForOrigin(state, origin) {
  return state.metamask.web3ShimUsageOrigins[origin];
}

/**
 * @typedef {Object} SwapsEthToken
 * @property {string} symbol - The symbol for ETH, namely "ETH"
 * @property {string} name - The name of the ETH currency, "Ether"
 * @property {string} address - A substitute address for the metaswap-api to
 * recognize the ETH token
 * @property {string} decimals - The number of ETH decimals, i.e. 18
 * @property {string} balance - The user's ETH balance in decimal wei, with a
 * precision of 4 decimal places
 * @property {string} string - The user's ETH balance in decimal ETH
 */

/**
 * Swaps related code uses token objects for various purposes. These objects
 * always have the following properties: `symbol`, `name`, `address`, and
 * `decimals`.
 *
 * When available for the current account, the objects can have `balance` and
 * `string` properties.
 * `balance` is the users token balance in decimal values, denominated in the
 * minimal token units (according to its decimals).
 * `string` is the token balance in a readable format, ready for rendering.
 *
 * Swaps treats the selected chain's currency as a token, and we use the token constants
 * in the SWAPS_CHAINID_DEFAULT_TOKEN_MAP to set the standard properties for
 * the token. The getSwapsDefaultToken selector extends that object with
 * `balance` and `string` values of the same type as in regular ERC-20 token
 * objects, per the above description.
 *
 * @param {object} state - the redux state object
 * @returns {SwapsEthToken} The token object representation of the currently
 * selected account's ETH balance, as expected by the Swaps API.
 */

export function getSwapsDefaultToken(state) {
  const selectedAccount = getSelectedAccount(state);
  const { balance } = selectedAccount;
  const chainId = getCurrentChainId(state);

  const defaultTokenObject = SWAPS_CHAINID_DEFAULT_TOKEN_MAP[chainId];

  return {
    ...defaultTokenObject,
    balance: hexToDecimal(balance),
    string: getValueFromWeiHex({
      value: balance,
      numberOfDecimals: 4,
      toDenomination: 'ETH',
    }),
  };
}

export function getIsSwapsChain(state) {
  const chainId = getCurrentChainId(state);
  return ALLOWED_SWAPS_CHAIN_IDS[chainId];
}

export function getNativeCurrencyImage(state) {
  const nativeCurrency = getNativeCurrency(state).toUpperCase();
  return NATIVE_CURRENCY_TOKEN_IMAGE_MAP[nativeCurrency];
}

export function getNextSuggestedNonce(state) {
  return Number(state.metamask.nextNonce);
}

export function getShowWhatsNewPopup(state) {
  return state.appState.showWhatsNewPopup;
}

/**
 * Get an object of notification IDs and if they are allowed or not.
 * @param {Object} state
 * @returns {Object}
 */
function getAllowedNotificationIds(state) {
  return {
    1: true,
    2: true,
    3: true,
    4: getCurrentChainId(state) === BSC_CHAIN_ID,
    5: true,
    6: true,
  };
}

/**
 * @typedef {Object} Notification
 * @property {number} id - A unique identifier for the notification
 * @property {string} date - A date in YYYY-MM-DD format, identifying when the notification was first committed
 */

/**
 * Notifications are managed by the notification controller and referenced by
 * `state.metamask.notifications`. This function returns a list of notifications
 * the can be shown to the user. This list includes all notifications that do not
 * have a truthy `isShown` property.
 *
 * The returned notifications are sorted by date.
 *
 * @param {Object} state - the redux state object
 * @returns {Notification[]} An array of notifications that can be shown to the user
 */

export function getSortedNotificationsToShow(state) {
  const notifications = Object.values(state.metamask.notifications);
  const allowedNotificationIds = getAllowedNotificationIds(state);
  const notificationsToShow = notifications.filter(
    (notification) =>
      !notification.isShown && allowedNotificationIds[notification.id],
  );
  const notificationsSortedByDate = notificationsToShow.sort(
    (a, b) => new Date(b.date) - new Date(a.date),
  );
  return notificationsSortedByDate;
}

export function getShowRecoveryPhraseReminder(state) {
  const {
    recoveryPhraseReminderLastShown,
    recoveryPhraseReminderHasBeenShown,
  } = state.metamask;

  const currentTime = new Date().getTime();
  const frequency = recoveryPhraseReminderHasBeenShown ? DAY * 90 : DAY * 2;

  return currentTime - recoveryPhraseReminderLastShown >= frequency;
}

/**
 * To get the useTokenDetection flag which determines whether a static or dynamic token list is used
 * @param {*} state
 * @returns Boolean
 */
export function getUseTokenDetection(state) {
  return Boolean(state.metamask.useTokenDetection);
}

/**
 * To retrieve the tokenList produced by TokenListcontroller
 * @param {*} state
 * @returns {Object}
 */
export function getTokenList(state) {
  return state.metamask.tokenList;
}
