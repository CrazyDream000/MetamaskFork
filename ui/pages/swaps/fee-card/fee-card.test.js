import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { useSelector } from 'react-redux';

import { checkNetworkAndAccountSupports1559 } from '../../../selectors';
import {
  getGasEstimateType,
  getGasFeeEstimates,
  getIsGasEstimatesLoading,
} from '../../../ducks/metamask/metamask';

import {
  renderWithProvider,
  createSwapsMockStore,
  setBackgroundConnection,
  MOCKS,
} from '../../../../test/jest';
import { MAINNET_CHAIN_ID } from '../../../../shared/constants/network';
import FeeCard from '.';

const middleware = [thunk];

jest.mock('react-redux', () => {
  const actual = jest.requireActual('react-redux');

  return {
    ...actual,
    useSelector: jest.fn(),
  };
});

const generateUseSelectorRouter = () => (selector) => {
  if (selector === checkNetworkAndAccountSupports1559) {
    return true;
  }
  if (selector === getGasEstimateType) {
    return 'fee-market';
  }
  if (selector === getGasFeeEstimates) {
    return MOCKS.createGasFeeEstimatesForFeeMarket();
  }
  if (selector === getIsGasEstimatesLoading) {
    return false;
  }
  return undefined;
};

setBackgroundConnection({
  getGasFeeTimeEstimate: jest.fn(),
  getGasFeeEstimatesAndStartPolling: jest.fn(),
});

const createProps = (customProps = {}) => {
  return {
    primaryFee: {
      fee: '0.0441 ETH',
      maxFee: '0.04851 ETH',
    },
    secondaryFee: {
      fee: '$101.98',
      maxFee: '$112.17',
    },
    hideTokenApprovalRow: false,
    onFeeCardMaxRowClick: jest.fn(),
    tokenApprovalTextComponent: (
      <span
        key="swaps-view-quote-approve-symbol-1"
        className="view-quote__bold"
      >
        ABC
      </span>
    ),
    tokenApprovalSourceTokenSymbol: 'ABC',
    onTokenApprovalClick: jest.fn(),
    metaMaskFee: '0.875',
    isBestQuote: true,
    numberOfQuotes: 6,
    onQuotesClick: jest.fn(),
    tokenConversionRate: 0.015,
    chainId: MAINNET_CHAIN_ID,
    networkAndAccountSupports1559: false,
    ...customProps,
  };
};

describe('FeeCard', () => {
  it('renders the component with initial props', () => {
    useSelector.mockImplementation(generateUseSelectorRouter());
    const props = createProps();
    const { getByText } = renderWithProvider(<FeeCard {...props} />);
    expect(getByText('Using the best quote')).toBeInTheDocument();
    expect(getByText('6 quotes')).toBeInTheDocument();
    expect(getByText('Max network fee')).toBeInTheDocument();
    expect(getByText('Estimated network fee')).toBeInTheDocument();
    expect(getByText(props.primaryFee.fee)).toBeInTheDocument();
    expect(getByText(props.primaryFee.maxFee)).toBeInTheDocument();
    expect(getByText(props.secondaryFee.fee)).toBeInTheDocument();
    expect(getByText(props.secondaryFee.maxFee)).toBeInTheDocument();
    expect(
      getByText('Quote includes a 0.875% MetaMask fee'),
    ).toBeInTheDocument();
    expect(
      document.querySelector('.fee-card__savings-and-quotes-header'),
    ).toMatchSnapshot();
    expect(
      document.querySelector('.fee-card__top-bordered-row'),
    ).toMatchSnapshot();
  });

  it('renders the component with EIP-1559 enabled', () => {
    const store = configureMockStore(middleware)(createSwapsMockStore());
    const props = createProps({
      networkAndAccountSupports1559: true,
      maxPriorityFeePerGasDecGWEI: '3',
      maxFeePerGasDecGWEI: '4',
    });
    const { getByText } = renderWithProvider(<FeeCard {...props} />, store);
    expect(getByText('Using the best quote')).toBeInTheDocument();
    expect(getByText('6 quotes')).toBeInTheDocument();
    expect(getByText('Estimated gas fee')).toBeInTheDocument();
    expect(getByText('Maybe in 5 minutes')).toBeInTheDocument();
    expect(getByText(props.primaryFee.fee)).toBeInTheDocument();
    expect(getByText(props.secondaryFee.fee)).toBeInTheDocument();
    expect(getByText(`: ${props.secondaryFee.maxFee}`)).toBeInTheDocument();
    expect(
      getByText('Quote includes a 0.875% MetaMask fee'),
    ).toBeInTheDocument();
    expect(
      document.querySelector('.fee-card__savings-and-quotes-header'),
    ).toMatchSnapshot();
    expect(
      document.querySelector('.fee-card__top-bordered-row'),
    ).toMatchSnapshot();
  });
});
