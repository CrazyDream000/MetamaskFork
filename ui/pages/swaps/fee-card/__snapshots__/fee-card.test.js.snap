// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeeCard renders the component with EIP-1559 enabled 1`] = `
<div
  class="fee-card__savings-and-quotes-header"
  data-testid="fee-card__savings-and-quotes-header"
>
  <div
    class="fee-card__savings-and-quotes-row"
  >
    <p
      class="fee-card__savings-text"
    >
      Using the best quote
    </p>
    <div
      class="fee-card__quote-link-container"
    >
      <p
        class="fee-card__quote-link-text"
      >
        6 quotes
      </p>
      <div
        class="fee-card__caret-right"
      >
        <i
          class="fa fa-angle-up"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`FeeCard renders the component with EIP-1559 enabled 2`] = `
<div
  class="fee-card__top-bordered-row"
>
  <div
    class="fee-card__row-label"
  >
    <div
      class="fee-card__row-header-text"
    >
      Quote includes a 0.875% MetaMask fee
    </div>
    <div
      class="info-tooltip"
    >
      <div
        class="fee-card__info-tooltip-container"
      >
        <div
          aria-describedby="tippy-tooltip-7"
          class="info-tooltip__tooltip-container"
          data-original-title="null"
          data-tooltipped=""
          style="display: inline;"
          tabindex="0"
        >
          <svg
            viewBox="0 0 10 10"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 0C2.2 0 0 2.2 0 5s2.2 5 5 5 5-2.2 5-5-2.2-5-5-5zm0 2c.4 0 .7.3.7.7s-.3.7-.7.7-.7-.2-.7-.6.3-.8.7-.8zm.7 6H4.3V4.3h1.5V8z"
              fill="#b8b8b8"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`FeeCard renders the component with initial props 1`] = `
<div
  class="fee-card__savings-and-quotes-header"
  data-testid="fee-card__savings-and-quotes-header"
>
  <div
    class="fee-card__savings-and-quotes-row"
  >
    <p
      class="fee-card__savings-text"
    >
      Using the best quote
    </p>
    <div
      class="fee-card__quote-link-container"
    >
      <p
        class="fee-card__quote-link-text"
      >
        6 quotes
      </p>
      <div
        class="fee-card__caret-right"
      >
        <i
          class="fa fa-angle-up"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`FeeCard renders the component with initial props 2`] = `
<div
  class="fee-card__top-bordered-row"
>
  <div
    class="fee-card__row-label"
  >
    <div
      class="fee-card__row-header-text"
    >
      Quote includes a 0.875% MetaMask fee
    </div>
    <div
      class="info-tooltip"
    >
      <div
        class="fee-card__info-tooltip-container"
      >
        <div
          aria-describedby="tippy-tooltip-3"
          class="info-tooltip__tooltip-container"
          data-original-title="null"
          data-tooltipped=""
          style="display: inline;"
          tabindex="0"
        >
          <svg
            viewBox="0 0 10 10"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 0C2.2 0 0 2.2 0 5s2.2 5 5 5 5-2.2 5-5-2.2-5-5-5zm0 2c.4 0 .7.3.7.7s-.3.7-.7.7-.7-.2-.7-.6.3-.8.7-.8zm.7 6H4.3V4.3h1.5V8z"
              fill="#b8b8b8"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
